### PR TITLE
Initial build

### PR DESCRIPTION
**Pull request by stepanstipl**
Initial take on Locust image build. It follows the concept of the Locust chart we use, so it's compatible, swap-in replacement.

Notable differences:
- Adds Let's encrypt Staging certificates
- Uses latest version of Locust - 0.9.0
- Based on the latest Alpine 3.8
- Uses `locust` user (no root)
- No security issues (at least according to quay.io as of `22-01-2019`)
- Smaller size (`34 MB` instead of `42 MB`)

The new image has been tested using `rake test_flowmanager` and `rake test_preferences`.

### CODE REVIEW

**Comment by natarajaya** [comment]
LGTM!

> Adds Let's encrypt Staging certificates

I don't think let's encrypt root certs belong here, but I understand the reason why you added them – that's probably okay, if there is no other straightforward way to deal with Locust cert validation issue.

**Comment by mrtyler** [comment]
I'd like to take another pass at this but a couple things:
* It's unclear how much of this is copied from upstream vs new additions by you, which makes it harder to figure out what parts to focus on for review (e.g. how much do I need to worry about logic bugs in the Makefile)
* Adding the LE staging certs here seems like an okay workaround. However, I fear a scenario where we somehow deploy staging certs to production, but don't notice because our Locust smoke test passes. Maybe uptime checks would catch this scenario, and I'm being too paranoid? Might be worth a little more thought / discussion.

**Comment by stepanstipl** [comment]
thanks for reviews.

@natarajaya

> I don't think let's encrypt root certs belong here, but I understand the reason why you added them – that's probably okay, if there is no other straightforward way to deal with Locust cert validation issue.

I see your point. Originally I was thinking about making this parametrized (env variable) and before that coming with some approach that would allow adding certs via Helm/ConfigMap mechanism, but all of those would require filesystem changes during runtime and add more complexity.

Another approach would be disabling cert checking completely. But here I would argue that while we're not testing that the cert has been issued/not issued by particular CA, we want to test the functionality of the https/ssl/certs mechanism as such and therefore adding LE Staging certs is better than disabling certs check.

In the end, as this build is specific to our infrastructure, and we do use LE Staging certs in our environments, I decided to go with a simple solution.

@mrtyler 

> It's unclear how much of this is copied from upstream vs new additions by you, which makes it harder to figure out what parts to focus on for review (e.g. how much do I need to worry about logic bugs in the Makefile)

only `locust-tasks/tasks.py` is cp from upstream, rest is new.

> I fear a scenario where we somehow deploy staging certs to production, but don't notice because our Locust smoke test passes. Maybe uptime checks would catch this scenario

Stackdriver uptime checks would not catch this AFAIK (haven't tested but that's what the docs say here).

For me the question is "do we want to test for correct and valid certificates here?"  I don't think this is the purpose of the locust test. Otherwise it opens a lot of other questions (we still don't get expired certs, whole cert-manager mechanism does not have to be updated/completed when the locust tests run...).

I agree that we **should** monitor certificates, but IMHO that should be continuous process that catches things like expiring certs and I don't think doing it here is the right way. (One easy way to solve this is to add env variable switch, as mentioned above, but I particularly don't like that it would require fileystem modifications during runtime.)

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      py3-zmq \
+      python3 \
+    && apk add --no-cache --update --virtual build-dependencies \
+      gcc \
+      musl-dev \
+      python3-dev \
+    && pip3 install locustio=="${LOCUST_VERSION}" \
+    && mkdir -p "${LOCUST_HOME}" \
+    && addgroup -g "${LOCUST_GID}" "${LOCUST_GROUP}" \
+    && adduser -g "Kafka user" -D -h "${LOCUST_HOME}" -G "${LOCUST_GROUP}" -s /sbin/nologin -u "${LOCUST_UID}" "${LOCUST_USER}" \
```
Comment (amatas) [inline Dockerfile]: Kafka user? :laughing:

Comment (amatas) [reply Dockerfile]: Also Alpine comes with `nobody` and `guest` user, perhaps you can use one of them to save some lines of code. https://git.alpinelinux.org/aports/tree/main/alpine-baselayout/passwd#n27

**Review by amatas — APPROVED** [review]
LGTM

File: Dockerfile (hunk 1 above)
Comment (stepanstipl) [reply Dockerfile]: :D Ah cool, copy/paste, of course, will fix, thanks for catching this!

> nobody and guest user

I guess that would work as well, I prefer to create dedicated users for services to make things a bit more obvious, also helps when these need to own some files, but I guess that's just personal preference

**Comment by stepanstipl** [comment]
Thanks for reviews.

@mrtyler I've created an issue [GPII-3681
Monitor HTTPS certificates for validity & expiration](https://issues.gpii.net/browse/GPII-3681) to think about cert monitoring and I'm going to proceed with this. If there are any other outstanding concerns, let's pick them up afterwards.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      musl-dev \
+      python3-dev \
+    && pip3 install locustio=="${LOCUST_VERSION}" \
+    && mkdir -p "${LOCUST_HOME}" \
+    && addgroup -g "${LOCUST_GID}" "${LOCUST_GROUP}" \
+    && adduser -g "Locust user" -D -h "${LOCUST_HOME}" -G "${LOCUST_GROUP}" -s /sbin/nologin -u "${LOCUST_UID}" "${LOCUST_USER}" \
+    && chown -R "${LOCUST_USER}:${LOCUST_GROUP}" "${LOCUST_HOME}" \
+    && apk del build-dependencies \
+    && wget "${LE_STAGING_URL}" -O "${LE_STAGING_FILE}" \
+    && echo "${LE_STAGING_SHA}  ${LE_STAGING_FILE}" |  sha1sum -c -
```
Comment (mrtyler) [inline Dockerfile:30]: What is this for?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,40 @@
 # docker-image-locust
-Locust Docker image
+
+This images is based on Alpine and designed to be compatible with the Locust
```
Comment (mrtyler) [inline README.md:3]: image

File: README.md [hunk 4]
```diff
@@ -1,2 +1,40 @@
...
+chart we use. It also adds certificates for Let's Encrypt Staging, as we use
```
Comment (mrtyler) [inline README.md:4]: Who is "we"?

File: entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/bin/sh
+set -eo pipefail
+
+LOCUST_BIN='/usr/bin/locust'
```
Comment (mrtyler) [inline entrypoint.sh:4]: Double quotes for consistency

File: .gitlab-ci.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+stages:
+  - build
+
+build-push-master:
```
Comment (mrtyler) [inline .gitlab-ci.yaml:4]: It looks like we will only used tagged version of this image, so what is this step for?

File: entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+LOCUST_SCRIPT="${LOCUST_SCRIPT:-/locust-tasks/tasks.py}"
+LOCUST_MODE="${LOCUST_MODE:-standalone}"
+
+TARGET_HOST="${TARGET_HOST:?Required variable not set}"
+
+if [ "$LOCUST_MODE" == "master" ]; then
+  LOCUST_CMD="${LOCUST_BIN} -f ${LOCUST_SCRIPT} --host ${TARGET_HOST} --master"
+elif [[ "$LOCUST_MODE" = "worker" ]]; then
```
Comment (mrtyler) [inline entrypoint.sh:12]: Why does this branch use double brackets and single equals while the previous branch uses single brackets and double equals?

File: Dockerfile (hunk 1 above)
Comment (mrtyler) [reply Dockerfile]: I agree with a dedicated user vs overloading generic users.

**Comment by mrtyler** [comment]
I agree with separating responsibilities for performance/smoke testing (Locust) from responsibilities for validating TLS certs (some other facility).

File: Dockerfile (hunk 2 above)
Comment (stepanstipl) [reply Dockerfile:30]: This verifies that the downloaded certificate matches the SHA hash - it's a good practice to verify the integrity of any downloaded content when this is done manually, mostly security-wise (as opposed to using some packaging system that does this already, like `apk`)

File: README.md (hunk 3 above)
Comment (stepanstipl) [reply README.md:3]: good catch, thanks

File: README.md (hunk 4 above)
Comment (stepanstipl) [reply README.md:4]: Good point, I'll rephrase this to reference official Helm chart

File: entrypoint.sh (hunk 5 above)
Comment (stepanstipl) [reply entrypoint.sh:4]: There's a functional difference between single quotes and double quotes - no variable interpolation is done within single quotes. I prefer to use these accordingly (and believe it's a good practice)... and I admit I wasn't very good at doing this consistently in other places of this script.

File: .gitlab-ci.yaml (hunk 6 above)
Comment (stepanstipl) [reply .gitlab-ci.yaml:4]: While only tagged version of image should be used for gpii-infra, `latest` tag is also built on commits to master -> this is mostly for convenience and to keep `latest` tag in sync to latest code in `master`.

File: entrypoint.sh (hunk 7 above)
Comment (stepanstipl) [reply entrypoint.sh:12]: Good point, thanks for noticing this!

**Comment by stepanstipl** [comment]
@mrtyler Thanks a lot for reviewing this, some good points 👍  - I've addressed those in comments here and in https://github.com/gpii-ops/docker-image-locust/pull/2 PR, pls. have a quick look if it looks ok.

File: entrypoint.sh (hunk 5 above)
Comment (mrtyler) [reply entrypoint.sh:4]: Yes, in my experience the functional difference is this:

* (with double quotes) "I need to add a variable to this string. Done."

* (with single quotes) "I need to add a variable to this string. What, it didn't work? Oh, because I need to change it to double quotes."

For this reason, I prefer double quotes everywhere (except when they cannot be used). Most of our scripts (and most shell scripts on the internet) follow this pattern.

Comment (stepanstipl) [reply entrypoint.sh:4]: Hm, this is probably for longer discussion and highly opinionated. I agree with the argument for being consistent, but looking at our code you can find lot of things like:
- `num_failures=$(echo $SESSION_STATS | jq -r '.stats[] | select(.name == "Total").num_failures')`
- `-X POST -H 'Content-Type: application/json' -d '{"action": "finish_cluster"}')`
- `--user=$(gcloud info --format='value(config.account)')`

Anyway my thinking on how to improve this in long-term -> a) create a "style guide" to agree and clarify how this should look like and b) add some linter (popular one is https://github.com/koalaman/shellcheck) to our tests. What do you think?

Comment (mrtyler) [reply entrypoint.sh:4]: I agree this leads to a larger discussion. I agree that a style guide and a linter are the "real" solutions for this problem. However I think we generally use double quotes throughout the codebase pretty consistently.

The quoted examples look correct to me, because they are using single quotes to enable (unescaped) embedding of double quotes. (I agree that the last one should be double quotes unless there is some other consideration.)
